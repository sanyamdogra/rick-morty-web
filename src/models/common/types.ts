import { CharacterResponse, LocationResponse } from "../../services/types";

export interface NameLoc {
  name: string;
  url: string;
}

export interface Character {
  created: string;
  episode: string[];
  gender: string;
  id: number;
  image: string;
  location: NameLoc;
  name: string;
  origin: NameLoc;
  species: string;
  status: string;
  type: string;
  url: string;
}

export interface Episode {
  air_date: string;
  characters: string[];
  created: string;
  episode: string;
  id: number;
  name: string;
  url: string;
}

export type MockData =
  | Character
  | CharacterResponse
  | Episode
  | LocationResponse;
