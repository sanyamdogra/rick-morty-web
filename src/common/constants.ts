export const GITHUB_PROJECT_REPO =
  "https://github.com/sanyamdogra/rick-morty-web";
