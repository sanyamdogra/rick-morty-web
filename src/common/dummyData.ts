import { Character, Episode } from "../models/common/types";
import { CharacterResponse, LocationResponse } from "../services/types";

export const DUMMY_CHARACTER: Character = {
  created: "2017-11-04T18:48:46.250Z",
  episode: [
    "https://rickandmortyapi.com/api/episode/1",
    "https://rickandmortyapi.com/api/episode/2",
  ],
  gender: "Male",
  id: 1,
  image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
  location: {
    name: "Citadel of Ricks",
    url: "https://rickandmortyapi.com/api/location/3",
  },
  name: "Rick Sanchez",
  origin: {
    name: "Earth (C-137)",
    url: "https://rickandmortyapi.com/api/location/1",
  },
  species: "Human",
  status: "Alive",
  type: "",
  url: "https://rickandmortyapi.com/api/character/1",
};

export const DUMMY_EPISODES: Episode[] = [
  {
    air_date: "December 2, 2013",
    characters: [
      "https://rickandmortyapi.com/api/character/1",
      "https://rickandmortyapi.com/api/character/2",
    ],
    created: "2017-11-10T12:56:33.798Z",
    episode: "S01E01",
    id: 1,
    name: "Pilot",
    url: "https://rickandmortyapi.com/api/episode/1",
  },
  {
    air_date: "December 2, 2013",
    characters: [
      "https://rickandmortyapi.com/api/character/1",
      "https://rickandmortyapi.com/api/character/2",
    ],
    created: "2017-11-10T12:56:33.798Z",
    episode: "S01E01",
    id: 1,
    name: "Pilot",
    url: "https://rickandmortyapi.com/api/episode/1",
  },
];

export const DUMMY_EPISODE: Episode = {
  air_date: "December 2, 2013",
  characters: [
    "https://rickandmortyapi.com/api/character/1",
    "https://rickandmortyapi.com/api/character/2",
  ],
  created: "2017-11-10T12:56:33.798Z",
  episode: "S01E01",
  id: 1,
  name: "Pilot",
  url: "https://rickandmortyapi.com/api/episode/1",
};

export const DUMMY_LOCATION: LocationResponse = {
  created: "2017-11-10T13:08:13.191Z",
  dimension: "unknown",
  id: 3,
  name: "Citadel of Ricks",
  residents: [
    "https://rickandmortyapi.com/api/character/8",
    "https://rickandmortyapi.com/api/character/14",
  ],
  type: "Space station",
  url: "https://rickandmortyapi.com/api/location/3",
};

export const DUMMY_CHARACTER_RESPONSE: CharacterResponse = {
  info: {
    count: 826,
    next: "https://rickandmortyapi.com/api/character?page=2",
    pages: 42,
    prev: null,
  },
  results: [DUMMY_CHARACTER],
};
