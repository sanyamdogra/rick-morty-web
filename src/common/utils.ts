import { MockData } from "../models/common/types";

export const getEpisodeParams = (arr: string[]) => {
  return arr.reduce((acc: number[], str: string) => {
    const strArr = str.split("/");
    acc.push(parseInt(strArr[strArr.length - 1]));

    return acc;
  }, []);
};

export const mockFetchResponse = (data: MockData) => {
  global.fetch = jest.fn().mockImplementationOnce(() =>
    Promise.resolve({
      json: () => Promise.resolve(data),
    })
  );
};
