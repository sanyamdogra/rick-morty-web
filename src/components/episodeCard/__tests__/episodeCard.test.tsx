import React from "react";
import { render, screen, cleanup } from "@testing-library/react";
import EpisodeCard from "..";

import { DUMMY_EPISODES, DUMMY_EPISODE } from "../../../common/dummyData";

afterEach(cleanup);

test("should match the snapshot and render correct data", () => {
  const { asFragment } = render(
    <EpisodeCard episodeInfo={DUMMY_EPISODES} loading={false} />
  );

  expect(asFragment()).toMatchSnapshot();

  expect(screen.queryByTestId("episode-card-loader")).toBeNull();

  expect(screen.queryByTestId("card-field-value")).toHaveTextContent(
    "2 chapters"
  );

  expect(screen.getAllByTestId("episode-name")).toHaveLength(2);
});

test("render correct data for single episode object", () => {
  render(<EpisodeCard episodeInfo={DUMMY_EPISODE} loading={false} />);

  expect(screen.queryByTestId("episode-card-loader")).toBeNull();

  expect(screen.queryByTestId("card-field-value")).toHaveTextContent(
    "1 chapters"
  );

  expect(screen.getByTestId("episode-name")).toHaveTextContent("Pilot");
});

test("render correct data for loading state", () => {
  render(<EpisodeCard episodeInfo={DUMMY_EPISODE} loading={true} />);

  expect(screen.getByTestId("episode-card-loader")).toBeInTheDocument();

  expect(screen.queryByTestId("episode-list")).toBeNull();
});
