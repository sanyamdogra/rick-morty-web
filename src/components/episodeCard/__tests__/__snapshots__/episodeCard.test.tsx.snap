// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot and render correct data 1`] = `
<DocumentFragment>
  <div
    class="rick-morty-ep-card"
  >
    <div
      class="rick-morty-ep-card__heading"
    >
      Chapter list
    </div>
    <div
      class="rick-morty-card-field__label"
      data-testid="card-field-label"
    >
      Appears on
    </div>
    <div
      class="rick-morty-card-field__value rick-morty-card-field__value--none"
      data-testid="card-field-value"
    >
      2 chapters
    </div>
    <ul
      class="rick-morty-ep-card__ep-list"
      data-testid="episode-list"
    >
      <li
        data-testid="episode-name"
      >
        Pilot
      </li>
      <li
        data-testid="episode-name"
      >
        Pilot
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
